src/python/library/tritonclient/utils/cuda_shared_memory/_utils.py

```diff
@@ -0,0 +1,127 @@
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from cuda import cudart
+from typing import Any
+
+
+def _raise_errno_if_cuda_err(err, errno):
+    if isinstance(err, cudart.cudaError_t):
+        if (err != cudart.cudaError_t.cudaSuccess) and (errno != 0):
+            raise CudaSharedMemoryException(errno)
+
+
+def _raise_error(msg):
+    raise CudaSharedMemoryException(msg)
+
+
+class CudaSharedMemoryException(Exception):
+    """Exception indicating non-Success status.
+
+    Parameters
+    ----------
+    err : c_void_p
+        Pointer to an Error that should be used to initialize the exception.
+
+    """
+
+    def __init__(self, err):
+        self.err_code_map = {
+            -1:
+                "unable to set device successfully",
+            -2:
+                "unable to create cuda shared memory handle",
+            -3:
+                "unable to set values in cuda shared memory",
+            -4:
+                "unable to free GPU device memory",
+            -5:
+                "failed to read cuda shared memory results",
+            -6:
+                "unable to read device attributes",
+            -7:
+                "device or platform does not support unified virtual addressing",
+            -8:
+                "unable to manage CUDA stream"
+        }
+        self._msg = None
+        if type(err) == str:
+            self._msg = err
+        elif (err != 0) and (err in self.err_code_map):
+            self._msg = self.err_code_map[err]
+
+    def __str__(self):
+        msg = super().__str__() if self._msg is None else self._msg
+        return msg
+
+
+class CudaSharedMemoryHandle:
+
+    def __init__(self, triton_shm_name: str,
+                 cuda_shm_handle: cudart.cudaIpcMemHandle_t, base_addr: Any,
+                 byte_size: int, device_id: int) -> None:
+        self._triton_shm_name = triton_shm_name
+        self._cuda_shm_handle = cuda_shm_handle
+        self._base_addr = base_addr
+        self._byte_size = byte_size
+        self._device_id = device_id
+        # [FXIME] C implementation has below which is not relevant, need to
+        # revisit when applying similar change to system shared memory utils
+        # and check on whether the "handles" can be unified.
+        # handle->offset_ = 0;
+        # handle->shm_key_ = "";
+        # handle->shm_fd_ = 0;
+
+    def __del__(self):
+        # __init__() may fail, don't attempt to release
+        # uninitialized resource.
+        if not hasattr(self, "_base_addr"):
+            return
+        err, prev_device = cudart.cudaGetDevice()
+        _raise_errno_if_cuda_err(err, -1)
+        _raise_errno_if_cuda_err(cudart.cudaSetDevice(self._device_id), -1)
```

Comment:
@Tabrizian do you think that we can run into the issue of extra context creation?